samples/core/multi-device/main.c

```diff
@@ -0,0 +1,646 @@
+    cl_mem sub_input_grids[2] = { 0 };
+    cl_mem sub_output_grids[2] = { 0 };
+    cl_event events[2] = { 0 };
+
+    // Parse command-line options.
+    struct cl_sdk_options_Diagnostic diag_opts = { .quiet = false,
+                                                   .verbose = false };
+
+    // By default assume that there is only one device available.
+    // dev_opts->number is set to 1 so that when calling to cl_util_get_device
+    // for the second device there is no index out of range.
+    struct cl_sdk_options_SingleDevice dev_opts = {
+        .triplet = { 0, 0, CL_DEVICE_TYPE_ALL }
+    };
+    struct convolution_options convolution_opts = { .x_dim = 4096,
+                                                    .y_dim = 4096 };
+
+    OCLERROR_RET(
+        parse_options(argc, argv, &diag_opts, &dev_opts, &convolution_opts),
+        error, end);
+
+    // Create runtime objects based on user preference or default.
+    OCLERROR_PAR(dev = cl_util_get_device(dev_opts.triplet.plat_index,
+                                          dev_opts.triplet.dev_index,
+                                          dev_opts.triplet.dev_type, &error),
+                 error, dev);
+
+    if (!diag_opts.quiet)
+    {
+        cl_util_print_device_info(dev);
+    }
+
+    if (diag_opts.verbose)
+    {
+        printf("Creating sub-devices...");
+        fflush(stdout);
+    }
+
+#if CL_HPP_TARGET_OPENCL_VERSION < 120
+    fprintf(stderr,
+            "Error: OpenCL subdevices not supported before version 1.2 ");
+    exit(EXIT_FAILURE);
+#endif
+
+    // Create subdevices, each with half of the compute units available.
+    cl_uint max_compute_units = 0;
+    OCLERROR_RET(clGetDeviceInfo(dev, CL_DEVICE_MAX_COMPUTE_UNITS,
+                                 sizeof(cl_uint), &max_compute_units, NULL),
+                 error, dev);
+    cl_device_partition_property subdevices_properties[] = {
+        (cl_device_partition_property)CL_DEVICE_PARTITION_EQUALLY,
+        (cl_device_partition_property)(max_compute_units / 2), 0
+    };
+
+    // Initialize subdevices array with one device and then reallocate for
+    // MacOS and Windows not to complain about NULL subdevices array.
+    cl_uint subdev_count = 1;
+    cl_device_id* subdevices =
+        (cl_device_id*)malloc(subdev_count * sizeof(cl_device_id));
+
+    OCLERROR_RET(clCreateSubDevices(dev, subdevices_properties,
+                                    max_compute_units, subdevices,
+                                    &subdev_count),
+                 error, dev);
+
+    if (subdev_count < 2)
+    {
+        fprintf(stderr, "Error: OpenCL cannot create subdevices");
+        exit(EXIT_FAILURE);
+    }
+
+    subdevices =
+        (cl_device_id*)realloc(subdevices, subdev_count * sizeof(cl_device_id));
+    OCLERROR_RET(clCreateSubDevices(dev, subdevices_properties, subdev_count,
+                                    subdevices, NULL),
+                 error, subdevs);
```

Comment:
So what is happening here is that we need to do two calls to `clCreateSubDevices`: one for getting the value of the number of subdevices that will be created (this call has the `subdevices` array pointer as NULL) and another one for actually partitioning the devices into the sub-devices. 

This procedure would usually look like this:
```C
cl_uint subdev_count = 0;

// Query number of subdevices to be created.
clCreateSubDevices(dev, subdevices_properties, max_compute_units, NULL, &subdev_count);

// Allocate memory needed.
cl_device_id* subdevices = (cl_device_id*)malloc(subdev_count * sizeof(cl_device_id));

// Create subdevices.
clCreateSubDevices(dev, subdevices_properties, subdev_count, subdevices, NULL);
```

But `clCreateSubDevices`, when on Windows or MacOS, expects to receive a non-null pointer to the `subdevices` array. So what it was done to work around this is the following:
- initialize the `subdevices` array with size 1
- do the first call to `clCreateSubDevices` to get the value of `subdev_count`
- reallocate the memory reserved for the `subdevices` array now taking into account the correct value of `subdev_count`
- perform the second call to `clCreateSubDevices` to actually create the sub-devices

Reply:
> one for getting the value of the number of subdevices that will be created (this call has the `subdevices` array pointer as NULL)

Ah, the problem here though is that the `subdevices` array pointer is not NULL, since it was malloc'd immediately before the first call to `clCreateSubDevices`.  This means that two sets of sub-devices are created.

It would probably be OK if the `subdevices` array pointer were NULL for the first call.

Reply:
Yes, what I meant is that if we set the first array as NULL then on windows and mac that call fails (cause the `clCreateSubDevices` implementations expect non-null arrays). Is there any other (better) way to work around this rather than the one that is currently on the example?

Reply:
I think this is the best option:

Since we're partitioning equally into sub-devices with `MAX_COMPUTE_UNITS / 2` compute units, the number of sub-devices must be two.  This means that we should be able to avoid the dynamic allocation entirely, and just use a statically allocated array and one call to clCreateSubDevices.  Something like:

```c
    cl_uint subdev_count = 1;
    cl_device_id subdevices[2];
    OCLERROR_RET(clCreateSubDevices(dev, subdevices_properties,
                                    2, subdevices,
                                    &subdev_count),
                 error, props);
```

Other things we could consider are:

1. Dynamically allocate a really big array, say based on MAX_COMPUTE_UNITS, and just use that instead.  Some of the entries in the array (all but the first two) will be unassigned, but oh well.
2. Do the right thing and make the first call with a NULL sub-device pointer to get the sub-device count (which should be two, unconditionally), then use it for the dynamic allocation, even though this doesn't work for some implementations (which ones? do we have a CTS gap?).

Reply:
I like this approach, I think we can set a const variable `n` with the number of subdevices to be created, and then partition the device giving `MAX_COMPUTE_UNITS / n`. This logic is now implemented in the C sample.

Reply:
Also, regarding the second point mentioned, AFAIK this is not a CTS issue but a spec issue (the spec says nothing about what should happen in this situation).